File: 23/py/09.py
from itertools import pairwise
from typing import List
from toolz import pipe, curry, first, last


def parse_line(line: str):
    return pipe(
        line,
        lambda s: s.replace("\n", ""),
        curry(str.split, sep = " "),
        # curry(filter, str.isdigit),
        curry(map, int),
        list
    )


def diff(series: List[int]):
    return [b - a for a, b in pairwise(series)]


def recursive_extrapolation(series: List[int]) -> int:
    zeros = [0] * len(series)
    if series == zeros:
        return last(series)
    else:
        cur = last(series)
        lower = recursive_extrapolation(diff(series))
        return cur + lower


def iterative_extrapolation(series: List[int]) -> int:
    m = series.copy()
    lasts = []
    while m != ([0] * len(m)):
        lasts.append(m[-1])
        m = diff(m)
    return m[-1] + sum(lasts)


if __name__ == "__main__":

    file = "23/data/09/final.txt"
    lines = open(file).readlines()
    series = first(map(parse_line, lines))

    pipe(
        lines,
        curry(map, parse_line),
        curry(map, recursive_extrapolation),
        sum
    )

    pipe(
        lines,
        curry(map, parse_line),
        curry(map, iterative_extrapolation),
        sum
    )

File: 23/jl/09.jl
function convert_line(line::String)::Vector{Int}
    line |> split .|> (s -> parse(Int, s))
end

#===============#
#=    first    =#
#===============#

function extend_history(hist::Vector{Int})::Int
    dh::Vector{Int} = diff(hist)
    if dh == zeros(length(dh))
        return last(hist)
    end
    return last(hist) + last(extend_history(dh))
end

function part_one(input::Vector{String})::Int
    histories::Vector{Vector{Int}} = convert_line.(input)
    return sum(extend_history.(histories))
end

@assert part_one(readlines("23/data/09/test.txt")) == 114
println("Part 1: ", part_one(readlines("23/data/09/final.txt")))

#================#
#=    second    =#
#================#

function part_two(input::Vector{String})
    histories::Vector{Vector{Int}} = convert_line.(input)
    return sum(extrapolate_backwards.(histories))
end

function extrapolate_backwards(hist::Vector{Int})
    dh = diff(hist)
    if dh == zeros(length(dh))
        return first(hist)
    end
    return first(hist) - first(extrapolate_backwards(dh))
end

@assert part_two(readlines("data/09/test.txt")) == 2
println("Part 2: ", part_two(readlines("data/09/final.txt")))

